hypothesis-python/src/hypothesis/strategies/_internal/numbers.py

```diff
@@ -486,60 +471,15 @@ def floats(
                 f"smallest negative normal {-smallest_normal}"
             )
 
-    # Any type hint silences mypy when we unpack these parameters
-    kw: Any = {"allow_subnormal": allow_subnormal, "width": width}
-    unbounded_floats = FloatStrategy(
-        allow_infinity=allow_infinity, allow_nan=allow_nan, **kw
-    )
-    if min_value is None and max_value is None:
-        return unbounded_floats
-    elif min_value is not None and max_value is not None:
-        if min_value == max_value:
-            assert isinstance(min_value, float)
-            result = just(min_value)
-        elif is_negative(min_value):
-            if is_negative(max_value):
-                return floats(min_value=-max_value, max_value=-min_value, **kw).map(
-                    operator.neg
-                )
-            else:
-                return floats(min_value=0.0, max_value=max_value, **kw) | floats(
-                    min_value=0.0, max_value=-min_value, **kw
-                ).map(
-                    operator.neg  # type: ignore
-                )
-        elif (
-            count_between_floats(min_value, max_value, width) > 1000
-            or not allow_subnormal
-        ):
-            return FixedBoundedFloatStrategy(
-                lower_bound=min_value, upper_bound=max_value, **kw
-            )
-        else:
-            ub_int = float_to_int(max_value, width)
-            lb_int = float_to_int(min_value, width)
-            assert lb_int <= ub_int
-            result = integers(min_value=lb_int, max_value=ub_int).map(
-                lambda x: int_to_float(x, width)
-            )
-    elif min_value is not None:
-        assert isinstance(min_value, float)
-        if is_negative(min_value):
-            # Ignore known bug https://github.com/python/mypy/issues/6697
-            return unbounded_floats.map(abs) | floats(  # type: ignore
-                min_value=min_value, max_value=-0.0, **kw
-            )
-        else:
-            result = unbounded_floats.map(lambda x: min_value + abs(x))
-    else:
-        assert isinstance(max_value, float)
-        if not is_negative(max_value):
-            return floats(
-                min_value=0.0, max_value=max_value, **kw
-            ) | unbounded_floats.map(lambda x: -abs(x))
-        else:
-            result = unbounded_floats.map(lambda x: max_value - abs(x))
+    min_value = min_value if min_value is not None else float("-inf")
+    max_value = max_value if max_value is not None else float("inf")
+    assert isinstance(min_value, float)
+    assert isinstance(max_value, float)
+    result: SearchStrategy = FloatStrategy(min_value, max_value, allow_nan=allow_nan)
 
+    if not allow_subnormal:
+        smallest = width_smallest_normals[width]
+        result = result.filter(lambda x: abs(x) >= smallest)
```

Comment:
I think this incorrectly excludes zeros?  In general I'd prefer to have an explicit `allow_subnormal` flag as part of the `FloatsStrategy`; if False then the only allowed mantissa for the minimum exponent (iirc `=0` but check) is also exactly zero.  Conveniently I think we can handle this efficiently with your constrainer!

Reply:
Right. Made these changes, PTAL!